File: env/path.go
/*
 * Copyright (c) NeXTHub Corporation. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 */

package env

import (
	"log"
	"os"
	"path/filepath"
	"syscall"
)

var (
	// This is set by the linker.
	defaultGopRoot string
)

// GOPROOT returns the root of the Go+ tree. It uses the GOPROOT environment variable,
// if set at process start, or else the root used during the Go+ build.
func GOPROOT() string {
	gopRoot, err := findGopRoot()
	if err != nil {
		log.Panicln("GOPROOT not found:", err)
	}
	return gopRoot
}

const (
	envGOPROOT = "GOPROOT"
)

func findGopRoot() (string, error) {
	envGopRoot := os.Getenv(envGOPROOT)
	if envGopRoot != "" {
		// GOPROOT must valid
		if isValidGopRoot(envGopRoot) {
			return envGopRoot, nil
		}
		log.Panicf("\n%s (%s) is not valid\n", envGOPROOT, envGopRoot)
	}

	// if parent directory is a valid gop root, use it
	exePath, err := executableRealPath()
	if err == nil {
		dir := filepath.Dir(exePath)
		parentDir := filepath.Dir(dir)
		if parentDir != dir && isValidGopRoot(parentDir) {
			return parentDir, nil
		}
	}

	// check defaultGopRoot, if it is valid, use it
	if defaultGopRoot != "" && isValidGopRoot(defaultGopRoot) {
		return defaultGopRoot, nil
	}

	// Compatible with old GOPROOT
	if home := HOME(); home != "" {
		gopRoot := filepath.Join(home, "gop")
		if isValidGopRoot(gopRoot) {
			return gopRoot, nil
		}
		goplusRoot := filepath.Join(home, "goplus")
		if isValidGopRoot(goplusRoot) {
			return goplusRoot, nil
		}
	}
	return "", syscall.ENOENT
}

// Mockable for testing.
var executable = func() (string, error) {
	return os.Executable()
}

func executableRealPath() (path string, err error) {
	path, err = executable()
	if err == nil {
		path, err = filepath.EvalSymlinks(path)
		if err == nil {
			path, err = filepath.Abs(path)
		}
	}
	return
}

func isFileExists(path string) bool {
	st, err := os.Stat(path)
	return err == nil && !st.IsDir()
}

func isDirExists(path string) bool {
	st, err := os.Stat(path)
	return err == nil && st.IsDir()
}

func isValidGopRoot(path string) bool {
	return isDirExists(filepath.Join(path, "cmd/gop")) && isFileExists(filepath.Join(path, "go.mod"))
}

File: stdlib/public/LSP/IndexStoreDB/IndexStoreDB_Support/Path.cpp
//===--- Path.cpp ---------------------------------------------------------===//
//
// Copyright (c) NeXTHub Corporation. All rights reserved.
// DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
//
// This code is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// version 2 for more details (a copy is included in the LICENSE file that
// accompanied this code).
//
// Author(-s): Tunjay Akbarli
//
//===----------------------------------------------------------------------===//

#include <IndexStoreDB_Support/Path.h>
#include <IndexStoreDB_LLVMSupport/llvm_ADT_StringMap.h>
#include <IndexStoreDB_LLVMSupport/llvm_Support_Allocator.h>
#include <IndexStoreDB_LLVMSupport/llvm_Support_Path.h>
#include <IndexStoreDB_LLVMSupport/llvm_Support_FileSystem.h>
#include <IndexStoreDB_LLVMSupport/llvm_Support_Mutex.h>
#if defined(_WIN32)
#include <Windows.h>
#else
#include <unistd.h>
#endif
#include <limits.h>
#include <stdlib.h>

#if defined(_WIN32)
#define PATH_MAX MAX_PATH
#endif

using namespace IndexStoreDB;

namespace {
class CanonicalPathCacheImpl {
  llvm::StringMap<CanonicalFilePathRef, llvm::BumpPtrAllocator> CanonPaths;
  mutable llvm::sys::Mutex StateMtx;

public:
  CanonicalFilePath getCanonicalPath(StringRef Path,
                                     StringRef WorkingDir = StringRef());
};
}

CanonicalFilePath
CanonicalPathCacheImpl::getCanonicalPath(StringRef Path, StringRef WorkingDir) {
  if (Path.empty())
    return CanonicalFilePath();

  SmallString<256> AbsPath;
  if (llvm::sys::path::is_absolute(Path)) {
    AbsPath = Path;
  } else {
    assert(!WorkingDir.empty() && "passed relative path without working-dir");
    AbsPath = WorkingDir;
    AbsPath += '/';
    AbsPath += Path;
  }

  {
    llvm::sys::ScopedLock L(StateMtx);
    auto It = CanonPaths.find(AbsPath);
    if (It != CanonPaths.end())
      return It->second;
  }

  llvm::SmallString<PATH_MAX> Buffer;
  if (llvm::sys::fs::real_path(AbsPath.c_str(), Buffer, false)) {
    return CanonicalFilePathRef::getAsCanonicalPath(AbsPath);
  }
  StringRef CanonPath = Buffer;

  {
    llvm::sys::ScopedLock L(StateMtx);
    auto Pair = CanonPaths.insert(std::make_pair(AbsPath.str(), CanonicalFilePathRef()));
    auto &It = Pair.first;
    bool WasInserted = Pair.second;
    if (!WasInserted)
      return It->second;

    CanonicalFilePathRef CanonPathRef;
    if (CanonPath == It->first()) {
      CanonPathRef = CanonicalFilePathRef::getAsCanonicalPath(It->first());
    } else {
      auto &Alloc = CanonPaths.getAllocator();
      char *CopyPtr = Alloc.Allocate<char>(CanonPath.size());
      std::uninitialized_copy(CanonPath.begin(), CanonPath.end(), CopyPtr);
      StringRef CopyCanonPath(CopyPtr, CanonPath.size());
      CanonPathRef = CanonicalFilePathRef::getAsCanonicalPath(CopyCanonPath);
    }
    It->second = CanonPathRef;
    return CanonPathRef;
  }
}


CanonicalPathCache::CanonicalPathCache() {
  Impl = new CanonicalPathCacheImpl();
}
CanonicalPathCache::~CanonicalPathCache() {
  delete static_cast<CanonicalPathCacheImpl*>(Impl);
}

CanonicalFilePath
CanonicalPathCache::getCanonicalPath(StringRef Path, StringRef WorkingDir) {
  return static_cast<CanonicalPathCacheImpl*>(Impl)->getCanonicalPath(Path, WorkingDir);
}
